homeassistant/components/mqtt/remote.py

```diff
@@ -0,0 +1,237 @@
+from homeassistant.components.remote import RemoteDevice
+from homeassistant.const import (
+    CONF_COMMAND,
+    CONF_DEVICE,
+    CONF_ICON,
+    CONF_NAME,
+    CONF_OPTIMISTIC,
+    CONF_PAYLOAD_OFF,
+    CONF_PAYLOAD_ON,
+    CONF_VALUE_TEMPLATE,
+)
+from homeassistant.core import callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.restore_state import RestoreEntity
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+from . import (
+    CONF_COMMAND_TOPIC,
+    CONF_QOS,
+    CONF_RETAIN,
+    CONF_STATE_TOPIC,
+    MqttAttributes,
+    MqttAvailability,
+    MqttEntityDeviceInfo,
+    subscription,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "MQTT Remote"
+DEFAULT_OPTIMISTIC = False
+DEFAULT_PAYLOAD_ON = "ON"
+DEFAULT_PAYLOAD_OFF = "OFF"
+CONF_STATE_ON = "state_on"
+CONF_STATE_OFF = "state_off"
+CONF_COMMANDS = "commands"
+
+COMMAND_SCHEMA = vol.Schema(
+    {vol.Required(CONF_COMMAND): vol.All(str, vol.Length(min=1))}
+)
+
+PLATFORM_SCHEMA = (
+    mqtt.MQTT_RW_PLATFORM_SCHEMA.extend(
+        {
+            vol.Optional(CONF_DEVICE): mqtt.MQTT_ENTITY_DEVICE_INFO_SCHEMA,
+            vol.Optional(CONF_ICON): cv.icon,
+            vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+            vol.Optional(CONF_OPTIMISTIC, default=DEFAULT_OPTIMISTIC): cv.boolean,
+            vol.Optional(CONF_PAYLOAD_OFF, default=DEFAULT_PAYLOAD_OFF): cv.string,
+            vol.Optional(CONF_PAYLOAD_ON, default=DEFAULT_PAYLOAD_ON): cv.string,
+            vol.Optional(CONF_STATE_TOPIC): mqtt.valid_subscribe_topic,
+            vol.Optional(CONF_STATE_OFF): cv.string,
+            vol.Optional(CONF_STATE_ON): cv.string,
+            vol.Optional(CONF_COMMANDS, default={}): cv.schema_with_slug_keys(
+                COMMAND_SCHEMA
+            ),
+        }
+    )
+    .extend(mqtt.MQTT_AVAILABILITY_SCHEMA.schema)
+    .extend(mqtt.MQTT_JSON_ATTRS_SCHEMA.schema)
+)
+
+
+async def async_setup_platform(
+    hass: HomeAssistantType, config: ConfigType, async_add_entities, discovery_info=None
+):
+    """Set up MQTT remote through configuration.yaml."""
+    await _async_setup_entity(config, async_add_entities, discovery_info)
+
+
+async def _async_setup_entity(config, async_add_entities, config_entry=None):
+    """Set up the MQTT remote."""
+    async_add_entities([MqttRemote(config, config_entry)])
+
+
+# pylint: disable=too-many-ancestors
+class MqttRemote(
+    MqttAttributes, MqttAvailability, MqttEntityDeviceInfo, RemoteDevice, RestoreEntity
+):
+    """Representation of a remote that can be controlled using MQTT."""
+
+    def __init__(self, config, config_entry):
+        """Initialize the MQTT remote."""
+        self._state = False
+        self._sub_state = None
+        self._optimistic = None
+
+        # Load config
+        self._config = config
+        self._optimistic = config[CONF_OPTIMISTIC]
+        self._commands = config.get(CONF_COMMANDS)
```

Comment:
Please use `dict[key]` for required config keys and keys with default config schema values.